### DISCUSSION TITLE
Programmatic External Control (and introspection): for example how to set a variable externally

### DISCUSSION THREAD

**Discussion by yuki-tsubaki** [qtile/qtile#5033] [Q&A] [answered]
Qtile is generally great but I have always been puzzled because of the seeming distance between internals and access methods (which may just be on me for not understanding it well--I don't know).

Anyway, I have often wanted to change things that don't have a function to modify them remotely (e.g. via  the `qtile cmd-obj` interface, the InteractiveCommandClient, and other such means).

### Example
To make this question more clear here is an example:
Say I want to programmatically change the margin (idk maybe to show more of the wallpaper as a non-forceful visual reminder to stop working at the end of the day[^1]).

Modifying the margin from in the config file is trivial (just set it in the layout). 

Doing so dynamically is more complicated (because you need to make special lazy functions and and use the proper object path depending on the exact method/decorator), but still not too disruptively difficult.

I can define a function:
```python
@lazy.layout.function
def change_layout_gaps(layout, adjustment):
    layout.margin = layout.margin + adjustment
    layout.cmd_reset()
``` 
And then call it for example using a key binding.

However I have never figured out how I would do a similar introspecting adjustment from any of the external command interfaces. Sure maybe I could bind everything I ever want to do to a key and emulate input, but this just seems wrong given this is already written in Python: it should be easy to do this kind of stuff.

### Question
Does anyone know what I am missing or is programmatic external control like this really just outside of Qtile's current abilities/scope? 

[^1]: Yes I know about hooks and stuff and I'm sure one could hack a funky function that is run on a hook that fires often that checks the time or something which would allow you to do this specific thing all in the config file. This approach avoids the root problem however: I want to use a tool suited to what I want to do instead of recreating a janky version of that tool in Qtile. For example, I want a single line in mcron and maybe a short Python script, not to re-implement an incomplete alternative to mcron in my Qtile config.

**Comment by ervinpopescu** [comment]
I think that you **can** use the `InteractiveCommandClient` for that:
```py
from libqtile.command.client import InteractiveCommandClient
c = InteractiveCommandClient()

def change_layout_gaps(layout, adjustment):
    layout.margin = layout.margin + adjustment
    layout.reset()

c.group.layout.function(change_layout_gaps, 5)
```

**BUT** for some reason I get:
```py
Traceback (most recent call last):
  File "/home/ervin/debug.py", line 11, in <module>
    c.group.layout.function(change_layout_gaps, 5)
  File "/usr/lib/python3.12/site-packages/libqtile/command/client.py", line 199, in __call__
    return self._command.execute(self._current_node, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/libqtile/command/interface.py", line 248, in execute
    status, result = self._client.send(
                     ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/libqtile/ipc.py", line 171, in send
    return asyncio.run(self.async_send(msg))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/libqtile/ipc.py", line 187, in async_send
    send_data = _IPC.pack(msg, is_json=self.is_json)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/libqtile/ipc.py", line 135, in pack
    msg_bytes = marshal.dumps(msg)
                ^^^^^^^^^^^^^^^^^^
ValueError: unmarshallable object
```

@elParaguayo might know more, I'll take a look to see if that's the correct way to call `function` on a `CommandObject`.

EDIT: `cmd_*` are deprecated, you can just use `reset` here.

**Reply by elParaguayo** [reply, marked as answer]
You can't send a function over IPC so you'd need to do something like:
```python
from libqtile.command.client import InteractiveCommandClient
c = InteractiveCommandClient()

def adjust_margin(adjustment):
    c.group.layout.eval(f"self.margin += {adjustment};self.group.layout_all()")
```

**Reply by ervinpopescu** [reply]
> You can't send a function over IPC

I suppose that's really the unmarshallable object the error is thrown for, but looks like [from python 3.13 we will get marshallable functions](https://docs.python.org/3/library/marshal.html#marshal.dumps) 🎉 

Buuut, your function still won't work, since eval will execute the code till `;`...

Huh, not even this will work:
```py
def change_layout_gaps(adjustment):
    c.group.layout.eval(f"self.margin += {adjustment}")
    c.group.layout.eval("self.group.layout_all()")
```

Any idea why? Does this take into account the fact that margin could be a list?

**Reply by elParaguayo** [reply]
The `eval` call should work as we pass the string to `exec` if `eval` fails.

If you print those eval statements in your code we should get an error message back.

**Reply by ervinpopescu** [reply]
> If you print those eval statements in your code we should get an error message back.

Was doing exactly that:
```py
from libqtile.command.client import InteractiveCommandClient

c = InteractiveCommandClient()

def change_layout_gaps(adjustment):
    print(c.group.layout.eval(f"self.margin += {adjustment}"))
    print(c.group.eval("self.layout_all()"))

change_layout_gaps(50)
```

```bash
$ python ~/debug.py
(True, None)
(True, 'None')
```

😿

**Reply by elParaguayo** [reply]
Ok. I'll need to test when I'm home tonight.

**Reply by ervinpopescu** [reply]
Probably something related to the `layout_all()`, I'd bet, will check this one out some more.

**Reply by yuki-tsubaki** [reply]
> > You can't send a function over IPC
> 
> I suppose that's really the unmarshallable object the error is thrown for, but looks like [from python 3.13 we will get marshallable functions](https://docs.python.org/3/library/marshal.html#marshal.dumps) 🎉
> 
> Buuut, your function still won't work, since eval will execute the code till `;`...
> 
> Huh, not even this will work:
> 
> ```python
> def change_layout_gaps(adjustment):
>     c.group.layout.eval(f"self.margin += {adjustment}")
>     c.group.layout.eval("self.group.layout_all()")
> ```
> 
> Any idea why? Does this take into account the fact that margin could be a list?

Interesting that you are having issues with this. I tried it and it seems to work fine (although I will still be glad if the more obvious solution starts working when we get Python 3.13).

I'm on Qtile version 0.28.2.dev0+gf1ed49bc.d20240813 with Python version 3.12.6.

Thanks to @ervinpopescu for the idea of using the eval command (I had been overlooking that).

Since I don't know if this works for everyone (as it apparently didn't for @ervinpopescu) I think I will keep this open/unanswered until we make sure that that option works generally (unless someone objects).

**Reply by ervinpopescu** [reply]
> Thanks to @ervinpopescu for the idea of using the eval command (I had been overlooking that).

I think you meant elParaguayo :).

> Since I don't know if this works for everyone (as it apparently didn't for @ervinpopescu) I think I will keep this open/unanswered until we make sure that that option works generally (unless someone objects).

Still need some more time to look into that `eval` behavior and see if i'm doing something wrong. Also my laptop's broken :(

**Comment by ervinpopescu** [comment]
It's working now, no idea how to reproduce, feel free to set @elParaguayo's response as the answer.
